### PR TITLE
Add support for setting mode field on file assets

### PR DESCRIPTION
**Pull request by yurrriq**
Add a `Mode` field to `FileAssetSpec` so the mode of a file asset can be specified, maintaining the current default of `0440`.

---

The motivating example for this is as follows:

```yaml
apiVersion: kops.k8s.io/v1alpha2
kind: Cluster
metadata:
  name: hello-world.local
# ...
spec:
  fileAssets:
  - content: |
      #! /usr/bin/env bash
      echo "Hello, world!"
    mode: "0550"
    name: hello-world
    path: /usr/local/bin/hello-world
  hooks:
  - manifest: |
      [Unit]
      Description=Say hello to the world

      [Service]
      Type=oneshot
      Restart=no
      ExecStart=/usr/local/bin/hello-world
    name: hello-world.service
    useRawManifest: true
# ...
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -464,7 +464,7 @@ apimachinery-codegen-client:
 		 --go-header-file "hack/boilerplate/boilerplate.generatego.txt"
 	go run k8s.io/code-generator/cmd/client-gen@${CODEGEN_VERSION} --v=0 --clientset-name="clientset" \
 		 --input-base=k8s.io/kops/pkg/apis --input-dirs=. --input="kops/,kops/v1alpha2,kops/v1alpha3" \
- 		 --output-package=k8s.io/kops/pkg/client/clientset_generated/  --output-base=$(TMPDIR) \
```
Comment (yurrriq) [inline Makefile:467]: Emacs complained about the stray ` ` here.

File: Makefile [hunk 2]
```diff
@@ -143,15 +143,15 @@ clean:
 
 .PHONY: codegen
 codegen:
-	go build -o ${KOPS_ROOT}/_output/bin k8s.io/kops/upup/tools/generators/...
+	go build -o ${KOPS_ROOT}/_output/bin/ k8s.io/kops/upup/tools/generators/...
```
Comment (yurrriq) [inline Makefile:146]: Without this change to the `codegen` and `verify-codegen` targets, on my system, I get the `fitask` binary at `${KOPS_ROOT}/_output/bin` instead of `${KOPS_ROOT}/_output/bin/fitask`.

**Comment by hakman** [comment]
@yurrriq Please update the commit message with the title @olemarkus set for the PR and use a similar approach for the Makefile commit. After that, it should be ok to merge. Thanks!

**Comment by yurrriq** [comment]
Reworded commits. Still need to test.

**Comment by yurrriq** [comment]
```
export KOPS_BASE_URL=https://storage.googleapis.com/kops-ci/pulls/pull-kops-e2e-kubernetes-aws/pull-70773d69c9
```

```yaml
spec:
  fileAssets:
  - content: |
      #! /usr/bin/env bash
      echo "We did it, Joe!"
    name: we-did-it
    mode: "0550"
    path: /usr/local/bin/we-did-it
  hooks:
  - manifest: |
      [Unit]
      Description=Prove that we did it

      [Service]
      Type=oneshot
      Restart=no
      ExecStart=/usr/local/bin/we-did-it
    name: we-did-it.service
    useRawManifest: true
```
```


```
ubuntu@ip-10-112-64-80:~$ systemctl status we-did-it.service
● we-did-it.service - Prove that we did it
     Loaded: loaded (/lib/systemd/system/we-did-it.service; static; vendor preset: enabled)
     Active: inactive (dead)

Jun 08 21:51:46 ip-10-112-64-80 systemd[1]: Starting Prove that we did it...
Jun 08 21:51:46 ip-10-112-64-80 we-did-it[4168]: We did it, Joe!
Jun 08 21:51:46 ip-10-112-64-80 systemd[1]: hello-world.service: Succeeded.
Jun 08 21:51:46 ip-10-112-64-80 systemd[1]: Finished Prove that we did it.
```

**Comment by yurrriq** [comment]
With kOps 1.23.2:

```yaml
spec:
  fileAssets:
  - content: |
      #! /usr/bin/env bash
      echo "We did it, Joe!"
    name: we-did-it
    path: /usr/local/bin/we-did-it
  hooks:
  - manifest: |
      [Unit]
      Description=Prove that we did it

      [Service]
      Type=oneshot
      Restart=no
      ExecStart=/usr/local/bin/we-did-it
    name: we-did-it.service
    useRawManifest: true
```

```
ubuntu@ip-10-112-64-119:~$ systemctl status we-did-it
● we-did-it.service - Prove that we did it
     Loaded: loaded (/lib/systemd/system/we-did-it.service; static; vendor preset: enabled)
     Active: failed (Result: exit-code) since Wed 2022-06-08 22:01:22 UTC; 9s ago
    Process: 5237 ExecStart=/usr/local/bin/we-did-it (code=exited, status=203/EXEC)
   Main PID: 5237 (code=exited, status=203/EXEC)

Jun 08 22:01:22 ip-10-112-64-119 systemd[1]: Starting Prove that we did it...
Jun 08 22:01:22 ip-10-112-64-119 systemd[1]: we-did-it.service: Main process exited, code=exited, status=>
Jun 08 22:01:22 ip-10-112-64-119 systemd[1]: we-did-it.service: Failed with result 'exit-code'.
Jun 08 22:01:22 ip-10-112-64-119 systemd[1]: Failed to start Prove that we did it.
```

**Comment by yurrriq** [comment]
tl;dr I have successfully validated this so we should be good to merge :tada:

**Comment by yurrriq** [comment]
I’ll plan to update (the source of) https://kops.sigs.k8s.io/cluster_spec/#fileassets today to mention mode.

**Review by hakman — APPROVED** [review]

**Comment by yurrriq** [comment]
Thanks, @olemarkus, for the guidance, and @hakman for reviewing and merging. :rocket:
